### PR TITLE
Load modified routes only in `LocalSearch::loadSolution`

### PR DESCRIPTION
**Pull request by leonlan**
This PR changes `LocalSearch::loadSolution` to only load routes that have been modified. This is particularly useful for #533 because ILS does not modify large parts of the solution between iterations and ILS runs many more iterations than HGS. 

Currently, loading solutions in ILS takes 15% of the total runtime for 1K client instances. This change brings that number down significantly. Furthermore, the change does not impact HGS' runtime because HGS does not run many iterations so loading solutions is not a bottleneck.

I profiled on CVRP/VRPTW 1K instances for both algorithms and here are the runtime percentages for load solution before and after change:

| Setting | Before change | After change |
|-----------|---------|--------|
| CVRP ILS | 14.2% | 1.6% (65% time spent on updating modified routes) |
| VRPTW ILS | 15.2% | 3.0% (80% time spent on updating modified routes) |
| CVRP HGS | 0.1% | 0.1% |
| VRPTW HGS | 0.1% | 0.1% |

Time spent updating is the time spent on clearing/inserting/updating the modified routes. 

<details>

**Notes**:

Please read our [contributing guidelines](https://pyvrp.org/dev/contributing.html) first.
In particular:

- You must add tests when making code changes.
  This keeps the code coverage level up, and helps ensure the changes work as intended.
- When fixing a bug, you must add a test that would produce the bug in the master branch, and then show that it is fixed with the new code. 
- New code additions must be well formatted. Changes should pass the pre-commit workflow, which you can set up locally using [pre-commit](https://pre-commit.com/#intro). 
- Docstring additions must render correctly, including escapes and LaTeX.
- Finally, it is essential that all contributions in this PR are license-compatible with PyVRP's MIT license.
  Please check that this PR can be included into PyVRP under the MIT license.

</details>

### CODE REVIEW

**Comment by leonlan** [comment]
@N-Wouda could you review this? I don't know if we should merge this in main or if I just merge it into the `ils` branch first.

**Comment by N-Wouda** [comment]
> @N-Wouda could you review this? I don't know if we should merge this in main or just merge it into the `ils` branch first.

I think we can merge this into `main`. Could you profile this again? It should hopefully be a little faster now.

File: pyvrp/cpp/search/LocalSearch.cpp [hunk 1]
```diff
@@ -390,20 +409,64 @@ void LocalSearch::loadSolution(Solution const &solution)
...
+                             : routes.begin() + vehicleOffset[vehicleType + 1];
+
+        auto const equal = [&](auto const &route) { return route == solRoute; };
+        auto const route = std::find_if(start, end, equal);
+        if (route != end)
+        {
+            isNewRoute[idx] = false;
+            keepExistingRoute[std::distance(routes.begin(), route)] = true;
+        }
+    }
```
Comment (N-Wouda) [inline pyvrp/cpp/search/LocalSearch.cpp:434]: What worries me is that this loop is quadratic in the number of vehicles. The other loops are all linear. Can we somehow make this faster too?

**Comment by leonlan** [comment]
> I think we can merge this into `main`. Could you profile this again? It should hopefully be a little faster now.

- VRPTW went from 3.0% -> 2.8% (80% update).
- CVRP went from 1.6% -> 2.3% (55% update).


The new implementation is still quadratic time and with homogeneous vehicle types but it should be slightly faster esp. with heterogeneous instances. I quickly ran MDVRPTW 1K instance which went from 1.3% -> 1.1%. It seems more future-proof with larger and heterogeneous fleet so I'm OK with merging this version.

**Comment by leonlan** [comment]
Superseded by #720. Closing this PR.

**Pull request closed without merging**